### PR TITLE
update AST to 9.1.1 & simplify AST initialization for spectral conversion

### PR DESCRIPTION
**Pull request by duidae**
this PR includes:
1. update AST to 9.1.1 
2. simplify AST initialization for spectral conversion

related to #787

### CODE REVIEW

File: wasm_src/ast_wrapper/ast_wrapper.cc [hunk 1]
```diff
@@ -80,129 +80,47 @@ EMSCRIPTEN_KEEPALIVE AstFrameSet* initFrame(const char* header)
...
+    AstFrameSet* found = static_cast<AstFrameSet*>astFindFrame(frameSet, spectralTemplate, " " );
+    if (!found)
     {
-        cout << "Some AST LIB error, check logs." << endl;
-        astClearStatus;
+        cout << "Spectral frame not found." << endl;
         return nullptr;
     }
-    else if (specframe == AST__NULL)
+    AstSpecFrame *specframe = static_cast<AstSpecFrame*>astGetFrame(found, AST__CURRENT);
```
Comment (veggiesaurus) [inline wasm_src/ast_wrapper/ast_wrapper.cc:89]: Why don't we rather load all the headers into a single frame set, and _then_ split it out into a spectral and spatial frame?

Comment (veggiesaurus) [reply wasm_src/ast_wrapper/ast_wrapper.cc:89]: For example, `initFrame` could init the entire `AstFrameSet`, and then `getSpatialFrame` and `getSpectralFrame` could pull out the appropriate frames

**Review by veggiesaurus — CHANGES_REQUESTED** [review]
This looks much better than our previous approach, but there's a problem with loading images with no WCS info (e.g. `LMCq.fits`). Seems to cause AST to crash.

**Comment by duidae** [comment]
kk, will look into it after vacation(next week)

**Comment by duidae** [comment]
using dev branch, the error also show up in AST ver. 8.7.1:

astGrid(Plot): No numerical labels can be produced for axis 1 using the supplied Frame format string 'd.*'

probably some changes between 8.6.3/8.7.1 about args for astPlot() Format(1)='d.\*', Format(2)='d.\*'

will try to fix it

**Comment by veggiesaurus** [comment]
> using dev branch, the error also show up in AST ver. 8.7.1:
> 
> astGrid(Plot): No numerical labels can be produced for axis 1 using the supplied Frame format string 'd.*'
> 
> probably some changes between 8.6.3/8.7.1 about args for astPlot() Format(1)='d._', Format(2)='d._'
> 
> will try to fix it

One possible solution would be to not apply the `Format` arguments when the WCS is invalid

**Comment by duidae** [comment]
There are 2 issues:
1.  in 8.6.3, AST initFrame() returns invalid after reading the headers of LMCq.fits, however, after ver. 8.7.1, it returns ok, so the dummy frame will not be created.

2. originally 'Format(1)=d.\*' is accepted in  8.6.3, but '\*' seems no longer available after 8.7.1

so a simple fix would change (2) by removing '*', and the axis will be like:

image

and I think for future refactoring, initWCS() & initFullWCS() could be integrated for simplicity.

@veggiesaurus do you think the solution above ok?

**Comment by veggiesaurus** [comment]
> There are 2 issues:
> 
> 1. in 8.6.3, AST initFrame() returns invalid after reading the headers of LMCq.fits, however, after ver. 8.7.1, it returns ok, so the dummy frame will not be created.
> 2. originally 'Format(1)=d.*' is accepted in  8.6.3, but '*' seems no longer available after 8.7.1
> 
> so a simple fix would change (2) by removing '*', and the axis will be like:
> 
> image
> 
> and I think for future refactoring, initWCS() & initFullWCS() could be integrated for simplicity.
> 
> @veggiesaurus do you think the solution above ok?

That sounds sensible. For (1), can we just check whether there's a spatial frame using AST? And if not, update the "validWCS" flag in FrameStore?

**Comment by veggiesaurus** [comment]
@duidae is this ready for a re-review yet?

**Comment by duidae** [comment]
@veggiesaurus  still working on spatial part, will notify you when ready, thanks.

File: src/components/StokesAnalysis/StokesAnalysisComponent.tsx [hunk 2]
```diff
@@ -187,16 +189,18 @@ export class StokesAnalysisComponent extends React.Component<WidgetProps> {
         if (data.length > 0 && frame && frame.channelInfo) {
             let channelInfo = frame.channelInfo;
             const zIndex = this.matchZindex(x, y, data);
-            let nearestIndex;
+            let nearestIndex = undefined;
```
Comment (veggiesaurus) [inline src/components/StokesAnalysis/StokesAnalysisComponent.tsx]: by default, variables are `undefined`

**Comment by duidae** [comment]
just tried David's approach to extract skyframe and replace original wcsinfo with skyframe, however it breaks the AST ticks, not sure what is going on, so I think this PR can only handle the spectral part, spatial part refactoring needs more investigation. (so far the spatial functionality  still works using previous implementation).

original:
image

simply replace wcsinfo with skyframe:
image

File: src/stores/FrameStore.ts [hunk 3]
```diff
@@ -880,6 +863,13 @@ export class FrameStore {
         }
     };
 
+    public invertToNativeWCS = (value: number): number => {
```
Comment (veggiesaurus) [inline src/stores/FrameStore.ts]: should this be "convert"?

Comment (duidae) [reply src/stores/FrameStore.ts]: Fixed

**Review by veggiesaurus — CHANGES_REQUESTED** [review]
I must have skimmed over your message before, because I see now that the issues for LMCq aren't actually fixed, since it's now just assigning the letter "d" to the ticks, instead of just displaying the ticks as pixel values, which it used to do. For images with no WCS, it should revert to pixel coords, like it used to.

Part of the issue might be that now the WCS is constructed correctly, so no dummy frame is initialised. Technically, it does have a valid transform (from pixels to some unknown type), so perhaps we should use that? 

@kswang1029 the header itself doesn't have a CTYPE or CDELT but casacore seems to be inserting a CDELT of 1.0 by default. Not really sure how to handle situations like this...

**Comment by duidae** [comment]
@veggiesaurus  tried a workaround for LMCq.fits  in commit b5d6e37b0be7eabd56e54db6365817d95612c017, please have a look to see if it's ok, thanks.

**Review by veggiesaurus — CHANGES_REQUESTED** [review]
I think that check might be too specific. I think there's a simper approach: If the FrameSet that AST generates has a `Domain` attribute of `"SKY"`, then it can be displayed with degrees or HMS. If not, then the coordinates should _not_ be formatted as degrees etc.

LMCq for example, gives a domain of `"AXIS1-AXIS2"`

**Comment by duidae** [comment]
changed to avoid sky-missing frame

**Review by veggiesaurus — APPROVED** [review]

**Pull request merged**